fl4health/parameter_exchange/parameter_selection_criteria.py

```diff
@@ -173,3 +176,43 @@ def smallest_increase_in_magnitude_scores(model: nn.Module, initial_model: Optio
         initial_tensor_values = initial_model_states[tensor_name]
         names_to_scores[tensor_name] = (-1) * (torch.abs(current_tensor_values) - torch.abs(initial_tensor_values))
     return names_to_scores
+
+
+def select_scores_and_sample_masks(model: nn.Module, initial_model: Optional[nn.Module]) -> Tuple[NDArrays, List[str]]:
+    """
+    Selection function that first selects the "weight_scores" and "bias_scores" parameters for the
+    masked layers, and then samples binary masks based on those scores to send to the server.
+    This function is meant to be used for the FedPM algorithm.
+    """
+    parameters_to_exchange_names = []
+    parameters_to_exchange = []
+    model_states = model.state_dict()
+    with torch.no_grad():
+        if is_masked_module(model):
+            parameters_to_exchange_names.append("weight_scores")
+            weight_scores = model_states["weight_scores"]
+            # Sample a mask from the (unbounded) probability scores.
+            weight_probs = torch.sigmoid(weight_scores).cpu().numpy()
+            weight_mask_sample = bernoulli.rvs(weight_probs)
```

Comment:
Because of the Bernoulli sampling performed here, the parameters selected are in fact binary masks. That is, even though their corresponding names are something like "weight_scores" or "bias_scores", the tensors are not actually scores. But later after the tensors have been aggregated by the strategy, they will become score tensors again. I allowed this misalignment because these parameter names will later be needed to load the model anyway.